server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java

```diff
@@ -0,0 +1,383 @@
+                        requestPeers();
+                    }
+                } else {
+                    logger.trace("{} no longer connected to {}", this, discoveryNode);
+                    removePeer();
+                }
+            }
+        }
+
+        void establishConnection() {
+            assert holdsLock() : "PeerFinder mutex not held";
+            assert getDiscoveryNode() == null : "unexpectedly connected to " + getDiscoveryNode();
+            assert active;
+
+            logger.trace("{} attempting connection", this);
+            transportAddressConnector.connectToRemoteMasterNode(transportAddress, new ActionListener<DiscoveryNode>() {
+                @Override
+                public void onResponse(DiscoveryNode remoteNode) {
+                    assert remoteNode.isMasterNode() : remoteNode + " is not master-eligible";
+                    assert remoteNode.equals(getLocalNode()) == false : remoteNode + " is the local node";
+                    synchronized (mutex) {
+                        if (active) {
+                            assert discoveryNode.get() == null : "discoveryNode unexpectedly already set to " + discoveryNode.get();
+                            discoveryNode.set(remoteNode);
+                            requestPeers();
+                        }
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.debug(() -> new ParameterizedMessage("{} connection failed", Peer.this), e);
+                    removePeer();
+                }
+            });
+        }
+
+        private void removePeer() {
+            final Peer removed = peersByAddress.remove(transportAddress);
+            assert removed == Peer.this;
+        }
+
+        private void requestPeers() {
+            assert holdsLock() : "PeerFinder mutex not held";
+            assert peersRequestInFlight == false : "PeersRequest already in flight";
+            assert active;
+
+            final DiscoveryNode discoveryNode = getDiscoveryNode();
+            assert discoveryNode != null : "cannot request peers without first connecting";
+
+            logger.trace("{} requesting peers from {}", this, discoveryNode);
+            peersRequestInFlight = true;
+
+            List<DiscoveryNode> knownNodes = getKnownPeers();
+
+            transportService.sendRequest(discoveryNode, REQUEST_PEERS_ACTION_NAME,
+                new PeersRequest(getLocalNode(), knownNodes),
+                new TransportResponseHandler<PeersResponse>() {
+
+                    @Override
+                    public PeersResponse read(StreamInput in) throws IOException {
+                        return new PeersResponse(in);
+                    }
+
+                    @Override
+                    public void handleResponse(PeersResponse response) {
+                        logger.trace("{} received {} from {}", Peer.this, response, discoveryNode);
```

Comment:
with discoveryNode added to Peer.toString, no need to add discoveryNode here.

Reply:
Sure, see e5414a170acd19535823c02537f6e68dd702b7a1